### PR TITLE
Circulation coherence review: audit trail, emails, queues, NCIP parity

### PR DESCRIPTION
**Pull request by fabiodalez-dev**
## Summary

Full coherence review of the reservation/loan system across five domains (states & transitions, queues & capacity, emails, audit trail, NCIP/mobile parity), followed by fixes for everything it surfaced — including an adversarial review pass on the complete diff. No schema changes, no migration needed.

### Audit trail
- Every autonomous sweep mutation now records a SYSTEM audit event: `loan.expired` (revoked pickups), `loan.overdue` (bulk flip, ids collected before the UPDATE), `reservation.expired`, `reservation.promoted`, `loan.lost`/`loan.damaged` on returns.
- `loadLoanSnapshot` includes `sanzione`; the DataIntegrity repair is no longer silent (`source=repair` events + borrower emails, application clock).

### Emails
- New `loan_picked_up` template (5 locales): the pickup was the last silent transition.
- Every motivo/fallback now renders in the **recipient's** locale via `translateInLocale` — including caller-supplied reasons, `Nessun addebito`, and NCIP cancellations.
- Empty-address guards fire before outbox persistence; the picked-up sender tolerates archived titles like its renewed sibling; the orphan `reservation_book_available` template is retired.

### Queues & capacity
- Capacity subquery is window-overlap aware: future reservation windows are no longer blocked by loans that will be gone by then.
- Admin reactivation re-checks physical copies, appends to the queue tail and triggers immediate promotion (parity with a fresh reservation).
- Post-expiry promotion failures no longer suppress committed expiry emails; the post-commit notification flush can no longer turn a durable update into an error response.

### NCIP / mobile parity
- RenewItem refuses date-overdue loans, resets recall state, emails the borrower; CheckOutItem stops sending a false "approved" email; audit events carry `source=ncip` through `LoanRepository::close($source)`.
- CheckInItem is deterministic on multi-loan titles: NCIP-priority class, any-origin fallback only when unambiguous, two same-class candidates are always refused instead of closing an arbitrary loan.

### Loans
- Lost/damaged returns release the copy preserving curated states; the penalty reaches CSV export and the detail view (configured `app.currency`, not a hardcoded €).
- Every deadline extension emails the borrower post-commit. The admin edit deliberately does not consume `numero_rinnovi` nor apply `max_renewals` (librarian override) — now documented in code.
- One archived-title convention: the user's own active/pending loans show the real title (matching mobile and email); bare requests on archived books stay hidden.

### Product decisions included
1. **Self-cancellation confirmation emails**: cancelling a reservation/request now sends a confirmation to the user.
2. **Auto-expiry of stale `pendente` rows**: the sweep now expires `pendente` rows with `origine IN ('prenotazione','richiesta','ncip')` whose window fully passed — previously `richiesta`/`ncip` rows dangled forever.

## Tests
- 5 new standalone behavioral suites (155 checks) against the real DB: sweep audit, NCIP coherence (incl. check-in ambiguity guard), mail coherence, queue/capacity coherence, loan outcome visibility.
- 5 existing suites realigned. Standalone battery: 165/165 suites green. PHPStan level 5 clean, 5-locale parity verified, full Playwright suite 137 passed / 8 skipped (install phase, app already installed).

## Summary by CodeRabbit

* **Nuove funzionalità**
  * Conferma via email del ritiro fisico e dell’avvio del prestito.
  * Notifiche localizzate per rinnovi, annullamenti, scadenze e copie perse o danneggiate.
  * Visualizzazione delle sanzioni nei dettagli del prestito e nei report CSV.
  * Prestiti e richieste restano visibili anche per titoli archiviati.
  * Promozione automatica delle code quando si libera disponibilità.

* **Correzioni**
  * Migliorata la gestione di copie perse, danneggiate o in manutenzione.
  * Rafforzati i controlli su rinnovi, capacità e prenotazioni.
  * Ridotti gli invii duplicati e gestita più chiaramente la manutenzione simultanea.

### CODE REVIEW

File: app/Controllers/UserActionsController.php [hunk 1]
```diff
@@ -21,37 +21,54 @@ public function reservationsPage(Request $request, Response $response, mysqli $d
...
+        // deve poterlo vedere e annullare, col titolo REALE (stessa
+        // convenzione di mobile ed email — CI-SOFT-DELETE-EXEMPT: impegno
+        // proprio dell'utente); le richieste nude su libri archiviati, che
+        // non impegnano nulla, restano nascoste dal predicato sotto.
         $sql = "SELECT pr.id, pr.libro_id, pr.data_prestito, pr.data_scadenza, pr.stato, pr.created_at,
                        l.titolo, l.copertina_url
                 FROM prestiti pr
-                JOIN libri l ON l.id = pr.libro_id
-                WHERE pr.utente_id = ? AND pr.stato = 'pendente' AND l.deleted_at IS NULL
+                LEFT JOIN libri l ON l.id = pr.libro_id
```
[the remark this answers is not in the document]
Comment (fabiodalez-dev) [reply app/Controllers/UserActionsController.php:35]: The three joins are deliberate, not omissions. Lines 35/62 implement the single archived-title convention for the user's OWN commitments (matching the mobile API and the email senders — the borrower receives overdue notices for that loan and must recognize it), and line 425 feeds the cancellation notice for the affected user. All three carry CI-SOFT-DELETE-EXEMPT markers and pass scripts/ci-check-soft-delete.py, which is the enforcement gate for the rule. Bare requests on archived books stay hidden via the predicate (fixed in 28bc733c, see the next thread).

File: app/Controllers/UserActionsController.php [hunk 2]
```diff
@@ -21,37 +21,54 @@ public function reservationsPage(Request $request, Response $response, mysqli $d
...
+                WHERE pr.utente_id = ? AND pr.stato = 'pendente'
+                  AND (l.id IS NOT NULL OR pr.copia_id IS NOT NULL)
```
[the remark this answers is not in the document]
Comment (fabiodalez-dev) [reply app/Controllers/UserActionsController.php]: Fixed in 28bc733c: the predicate now checks l.deleted_at IS NULL instead of l.id IS NOT NULL — after removing the join filter, l.id was always non-null and bare pending requests on archived books leaked back in.

File: app/Support/DataIntegrity.php [hunk 3]
```diff
@@ -1049,36 +1052,139 @@ public function fixDataInconsistencies(): array {
...
+                    $results['fixed']++;
+                    $repairAffectedBooks[(int) $dup['libro_id']] = true;
+                    ActivityLog::recordReservationEvent(
+                        $this->db,
+                        $dupId,
+                        'reservation.cancelled',
+                        $beforeSnapshot,
+                        source: 'repair',
+                        operatorId: ActivityLog::SYSTEM_OPERATOR
+                    );
```
[the remark this answers is not in the document]
Comment (fabiodalez-dev) [reply app/Support/DataIntegrity.php:1085]: I'm keeping the default action here on purpose. No call site in the codebase passes action: 'cancellazione' for these events — every cancellation path (admin, user self-service, sweep) records with 'aggiornamento', keeping the after snapshot. The reservation row still exists (stato=annullata), and the after snapshot is what preserves the attiva→annullata transition in the audit diff; 'cancellazione' empties it. The event name (reservation.cancelled/expired) is the semantic discriminator. Changing only these two sites would make the repair diverge from every other producer.

File: app/Support/DataIntegrity.php [hunk 4]
```diff
@@ -1161,6 +1267,28 @@ public function fixDataInconsistencies(): array {
...
+                    $repairReservationManager->flushDeferredNotifications();
+                } catch (\Throwable $flushErr) {
+                    SecureLogger::warning('Flush notifiche differite post-repair fallito', ['error' => $flushErr->getMessage()]);
+                }
+            }
+            if ($expiredQueueEmails !== []) {
+                $notificationService = new NotificationService($this->db);
+                foreach ($expiredQueueEmails as $expiredEmail) {
+                    try {
+                        $notificationService->sendQueueReservationExpiredNotification($expiredEmail['email'], $expiredEmail['variables']);
```
[the remark this answers is not in the document]
Comment (fabiodalez-dev) [reply app/Support/DataIntegrity.php]: Fixed in 28bc733c: a send=false return is now logged with the email hash, matching ReservationManager::flushDeferredNotifications().

File: app/Support/MaintenanceService.php [hunk 5]
```diff
@@ -547,6 +569,24 @@ public function repairInvalidReadyPickups(): int
...
+                // B4: la revoca del ritiro non deve essere silenziosa. Invio
+                // POST-commit (stesso pattern dei sender dei sweep gemelli):
+                // un errore di invio non deve entrare nel catch esterno.
+                try {
+                    // Chiave di catalogo GREZZA (non __()): il sender la
+                    // traduce nel locale del destinatario, non della sessione.
+                    (new NotificationService($this->db))->sendPickupCancelledNotification(
+                        $loanId,
+                        'La copia assegnata al tuo ritiro non è più disponibile. La tua prenotazione resta valida: riceverai un nuovo avviso appena una copia sarà pronta.'
+                    );
```
[the remark this answers is not in the document]
Comment (fabiodalez-dev) [reply app/Support/MaintenanceService.php:627]: Already satisfied: the exact key ('La copia assegnata al tuo ritiro non è più disponibile. La tua prenotazione resta valida: riceverai un nuovo avviso appena una copia sarà pronta.') is present and translated in all five catalogs (it/en/de/fr/da) — added in this same PR together with the sender change. translateInLocale resolves it in the recipient's locale.

File: app/Support/MaintenanceService.php [hunk 6]
```diff
@@ -1421,24 +1494,79 @@ public function updateOverdueLoans(): int
...
-            SET stato = 'in_ritardo'
-            WHERE stato = 'in_corso'
-            AND data_scadenza < ?
-            AND attivo = 1
-        ");
+        // Audit trail: gli id vanno raccolti (e lockati) PRIMA dell'UPDATE bulk,
+        // così ogni flip produce un evento loan.overdue con operatore SYSTEM
+        // dentro la stessa transazione della mutazione. Il FOR UPDATE serializza
+        // il set: nessuna riga può cambiare stato tra la SELECT e il flip.
+        $this->db->begin_transaction();
```
[the remark this answers is not in the document]
Comment (fabiodalez-dev) [reply app/Support/MaintenanceService.php]: Fixed in 28bc733c: updateOverdueLoans() now accepts insideTransaction (explicit caller-managed-transaction flag, same pattern as DataIntegrity::validateAndUpdateLoan) and skips begin/commit/rollback when set, rethrowing so the caller's rollback stays in charge. runAll() flow unchanged.

File: installer/database/data_fr_FR.sql [hunk 7]
```diff
@@ -354,6 +354,7 @@ INSERT IGNORE INTO `email_templates` VALUES (23,'loan_recall_notification','fr_F
...
+INSERT IGNORE INTO `email_templates` VALUES (27,'loan_picked_up','fr_FR','📖 Bonne lecture ! Votre prêt a commencé','<h2>Retrait confirmé</h2>\n<p>Bonjour {{utente_nome}},</p>\n<p>Vous avez retiré le livre et votre prêt a officiellement commencé :</p>\n<div style="background-color: #ecfdf5; padding: 15px; border-radius: 5px; margin: 20px 0; border-left: 4px solid #10b981;">\n    <p><strong>Livre :</strong> {{libro_titolo}}</p>\n    <p><strong>Date de retrait :</strong> {{data_prestito}}</p>\n    <p><strong>Date d\'échéance :</strong> {{data_scadenza}}</p>\n</div>\n<p><strong>Important :</strong> Pensez à rendre le livre avant la date d\'échéance. Vous recevrez un rappel quelques jours avant.</p>\n<p>Bonne lecture !</p>','Inviata quando il lettore ritira fisicamente il libro e il prestito inizia.',1,NULL,NULL);
```
[the remark this answers is not in the document]
Comment (fabiodalez-dev) [reply installer/database/data_fr_FR.sql]: Fixed in 28bc733c — and the same Italian description was in the en_US, de_DE and da_DK seeds too, so I localized all four.

File: storage/plugins/ncip-server/NcipServerPlugin.php [hunk 8]
```diff
@@ -2185,6 +2240,27 @@ private function cancelPendingNcipRequest(int $bookId, ?int $userId): array
...
+                    source: 'ncip'
+                );
+            } else {
+                \App\Support\ActivityLog::recordReservationEvent(
+                    $this->db,
+                    (int) $reservationId,
+                    'reservation.cancelled',
+                    $activityBefore,
+                    source: 'ncip'
+                );
```
[the remark this answers is not in the document]
Comment (fabiodalez-dev) [reply storage/plugins/ncip-server/NcipServerPlugin.php:2256]: Same convention call as in DataIntegrity: no producer in the codebase uses action: 'cancellazione' for loan/reservation cancellations — the rows persist with a terminal stato and the after snapshot preserves the transition. The NCIP path now matches LoanApprovalController and ReservationsAdminController exactly; making it the only site with 'cancellazione' semantics would be the inconsistency.

[the remark this answers is not in the document]
Comment (fabiodalez-dev) [reply storage/plugins/ncip-server/NcipServerPlugin.php:2256]: This is the pre-existing ActivityLog contract, unchanged by this PR (same literal on main): TYPES has no 'reservation' entry — 'loan' is the circulation family and meta.event (reservation.*) plus entity_id are the discriminators the feed and the tests already rely on. Introducing a new type would change the TYPES contract, feed filtering and the shape of historical rows, which is out of scope here; I'd rather do that, if ever, as its own change.

File: tests/circulation-sweep-audit.unit.php [hunk 9]
```diff
@@ -0,0 +1,450 @@
...
+        "INSERT INTO prenotazioni (libro_id, utente_id, data_inizio_richiesta, data_fine_richiesta, queue_position, stato)
+         VALUES (?, ?, ?, ?, 2, 'attiva')"
+    );
+    $stmt->bind_param('iiss', $bookQ, $userQ, $rqS, $rqE);
+    $stmt->execute();
+    $resQ2 = (int) $db->insert_id;
+    $stmt->close();
+} finally {
+    $db->query($trgSql);
+}
```
[the remark this answers is not in the document]
Comment (fabiodalez-dev) [reply tests/circulation-sweep-audit.unit.php:312]: Fixed in 0d13be7f: the recreation is also registered as a shutdown hook (idempotent — recreates only if the trigger is missing), so a fatal exit can no longer leave the test schema without the duplicate guard. The finally stays as the immediate restore.

File: tests/circulation-sweep-audit.unit.php [hunk 10]
```diff
@@ -0,0 +1,450 @@
...
+     VALUES (?, ?, ?, ?, 1, 'attiva')"
+);
+$stmt->bind_param('iiss', $bookX, $userX, $expiredAt, $expiredAt);
+$stmt->execute();
+$resX = (int) $db->insert_id;
+$stmt->close();
+
+$integrity = new DataIntegrity($db);
+$fixResult = $integrity->fixDataInconsistencies();
+$check($fixResult['errors'] === [] || $fixResult['errors'] === null, '17 fixDataInconsistencies completes without errors');
```
[the remark this answers is not in the document]
Comment (fabiodalez-dev) [reply tests/circulation-sweep-audit.unit.php]: Fixed in 0d13be7f with the suggested array_key_exists form.

File: tests/loan-archived-title-notifications.unit.php [hunk 11]
```diff
@@ -281,9 +281,13 @@ public function formatEmailDate(string $dateString, bool $includeTime = false, ?
     '24b its notification fetch keeps the standard soft-delete filter (consistent with the gate)');
 
 // ── 25-26. Recipient-locale reason fallback (rejected-loan emails) ───────────
-$check(substr_count($notifSrc, "translateInLocale('Nessun motivo specificato'") >= 2
+// 0.7.81: the loan-id based sendLoanRejectedNotification(int) was removed as
+// dead code, so exactly ONE reject sender (the Direct one, the only live path)
+// must localize the fallback reason in the recipient locale.
+$check(substr_count($notifSrc, "translateInLocale('Nessun motivo specificato'") >= 1
```
[the remark this answers is not in the document]
Comment (fabiodalez-dev) [reply tests/loan-archived-title-notifications.unit.php]: Fixed in 0d13be7f: the assertion is now anchored to the body of sendLoanRejectedNotificationDirect (signature to next function) and requires exactly one translateInLocale call there.

File: tests/mail-coherence-0781.unit.php [hunk 12]
```diff
@@ -0,0 +1,314 @@
...
+$stmt->bind_param('is', $bookId, $inventory);
+$stmt->execute();
+$copyId = (int) $db->insert_id;
+$stmt->close();
+
+$password = password_hash('MailCoherence!1', PASSWORD_DEFAULT);
+$card1 = 'ZZMC' . strtoupper($run);
+$stmt = $db->prepare(
+    "INSERT INTO utenti (codice_tessera, nome, cognome, email, password, tipo_utente, stato, email_verificata)
+     VALUES (?, 'NoMail', 'Coherence', '', ?, 'standard', 'attivo', 1)"
```
[the remark this answers is not in the document]
Comment (fabiodalez-dev) [reply tests/mail-coherence-0781.unit.php:201]: Fixed in 0d13be7f: the fixture persists locale=de_DE (different from the installation fallback) and check D expects translateInLocale('Nessun addebito', 'de_DE') directly, with an extra guard that the de_DE catalog actually translates the string — no more double-resolver tautology.

File: tests/mail-coherence-0781.unit.php [hunk 13]
```diff
@@ -0,0 +1,314 @@
...
+$stmt = $db->prepare("SELECT variables_json FROM email_delivery_outbox WHERE recipient_email = ? AND template_name = 'loan_copy_outcome' ORDER BY id DESC LIMIT 1");
+$stmt->bind_param('s', $emailWith);
+$stmt->execute();
+$row = $stmt->get_result()->fetch_assoc();
+$stmt->close();
+$vars = json_decode((string) ($row['variables_json'] ?? ''), true);
+// Separatore decimale dipendente dal locale del destinatario: 12,50 (it/de/fr)
+// oppure 12.50 (en) sono entrambi formati corretti dello stesso importo.
+$check(is_array($vars) && preg_match('/12[.,]50/', (string) ($vars['sanzione'] ?? '')) === 1,
+    'positive penalty still renders the formatted amount');
```
[the remark this answers is not in the document]
Comment (fabiodalez-dev) [reply tests/mail-coherence-0781.unit.php:290]: Fixed in 0d13be7f: the assertion now also requires the configured currency token (€ for EUR, ISO code otherwise, derived from app.currency like the sender does) next to the formatted amount.

File: tests/queue-capacity-coherence-0781.unit.php [hunk 14]
```diff
@@ -0,0 +1,598 @@
...
+ *    NO copie rows is rejected with capacity_full, like store();
+ *  - LoanMultiplicityPolicy dedup: the 3 request entry-points (POST
+ *    /user/loan, POST /user/reserve, POST /api/books/{id}/reserve) block a
+ *    duplicate title/user identically with the policy ON and OFF;
+ *  - reservationsPage shows an ACTIVE loan whose book was soft-deleted
+ *    (CI-SOFT-DELETE-EXEMPT join + title fallback).
+ *
+ * Dates derive from DateHelper::today() (application timezone).
+ *
+ * Run: php tests/queue-capacity-coherence-0781.unit.php
```
[the remark this answers is not in the document]
Comment (fabiodalez-dev) [reply tests/queue-capacity-coherence-0781.unit.php:35]: This is a standalone PHP unit suite (php tests/…), not a Playwright spec — the /tmp/run-e2e.sh wrapper applies to the E2E specs, and every other unit suite in tests/ documents the same php invocation. In 0d13be7f I extended the run doc to list the E2E_DB_* env overrides (with .env as the fallback) so credentials resolution is explicit, and fixed the stale header line about the removed title fallback.

File: app/Controllers/MaintenanceController.php [hunk 15]
```diff
@@ -99,8 +99,22 @@ public function performMaintenance(Request $request, Response $response, mysqli
...
+            // secondo, anche da due sessioni admin) eseguono lo sweep una sola
+            // volta, ma una ripetizione sequenziale del bottone riesegue
+            // sempre — l'admin che preme "esegui manutenzione" deve ottenere
+            // la manutenzione (con cooldown 1' le email del sweep sparivano
+            // per un minuto intero, visto dal vivo nella suite email E2E che
+            // preme il bottone quattro volte di fila).
+            // Il claim protegge SOLO la parte circolazione (sweep/email, la
+            // stessa del cron): la riparazione dati dei punti 2-3 — che
+            // runAll() NON copre — gira COMUNQUE, mai un successo a vuoto.
+            $circulation = (new MaintenanceService($db))->runIfNeeded(0);
```
[the remark this answers is not in the document]
Comment (fabiodalez-dev) [reply app/Controllers/MaintenanceController.php:106]: Fixed in 02b90cc0 + 14a488ce. runAll() now takes GET_LOCK('pinakes_maintenance_<db>', 0) for its entire execution and returns {skipped, reason: in_progress} when another connection holds it — that covers every entry point (admin button, login hook, cron) since they all funnel through runAll(), and it also protects the same-second window the claim used to leave open. I kept GET_LOCK instead of a lock table: it is connection-scoped and auto-released on connection death, so there is no stale-lock recovery to manage; release happens in a finally. The claim in runIfNeeded() stays for what it does well (cron/login cadence). The admin note no longer says 'try again in a minute': new message in all five catalogs. Behavioral coverage in circulation-sweep-audit checks 24-25 (second connection holds the lock -> skip; release -> runs again).

**Comment by fabiodalez-dev** [comment]
Re: the outside-diff suggestion to pass `action: 'cancellazione'` in the two NCIP audit calls (NcipServerPlugin.php ~2255-2261) — this is the same proposal I already declined in the earlier inline thread, and the situation is unchanged, so I'm declining it again for the same reason: no producer in the codebase passes `action: 'cancellazione'` for loan/reservation cancellations. The rows persist with a terminal stato (annullato/annullata) — they are not deleted — and keeping the after snapshot is what preserves the attiva→annullata transition in the audit diff; `'cancellazione'` empties it. The event name (loan.cancelled / reservation.cancelled) is the semantic discriminator. Making the NCIP path the only site with cancellation-action semantics would be the inconsistency, not the fix. If the convention ever changes, it should change for every producer at once, as its own scoped change.
